slackbridge/bots.py

```diff
@@ -198,28 +199,74 @@ class UserBot(IRCBot):
     def __init__(self, nickname, realname, user_id, channels,
                  target_group, nickserv_pw):
         self.nickname = '{}-slack'.format(utils.strip_nick(nickname))
+        self.intended_nickname = self.nickname
         self.realname = realname
         self.user_id = user_id
         self.channels = channels
         self.nickserv_password = nickserv_pw
         self.target_group_nick = target_group
 
+        self.nick_change_loop = LoopingCall(
+            self.setNick, self.intended_nickname)
+        self.signed_in = False
+
     def log(self, method, message):
         full_message = '[{}]: {}'.format(self.nickname, message)
         return method(full_message)
 
     def signedOn(self):
-        # If already registered, auth in
-        self.msg('NickServ', 'IDENTIFY {}'.format(self.nickserv_password))
-        # And if not, register for the first time
-        self.msg('NickServ', 'GROUP {} {}'.format(self.target_group_nick,
-                                                  self.nickserv_password))
+        self.nickserv_auth()
+
         for channel in self.channels:
             self.log(log.msg, 'Joining #{}'.format(channel['name']))
             self.join_channel(channel['name'])
 
         self.away('Default away for startup.')
 
+        # Needed for irc_ERR_NICKNAMEINUSE to tell if it's possible to use
+        # NickServ's RECOVER and RELEASE commands.
+        self.signed_in = True
+
+    def nickChanged(self, nick):
+        if nick != self.intended_nickname:
+            self.nick_change_loop.start(5)
+            self.log(log.msg,
+                     'Attempting to change nick to {}'.format(
+                         self.intended_nickname))
+        else:
+            if self.nick_change_loop.running:
+                self.nick_change_loop.stop()
+            self.log(log.msg,
+                     'Successfully changed nick to {}'.format(
+                         self.intended_nickname))
+
+    # Called when an attempted nickname change collides with an existing one.
+    def irc_ERR_NICKNAMEINUSE(self, prefix, params):
+        if not self.signed_in:
+            # Alter the nickname by appending '_' and sign in.
+            super().irc_ERR_NICKNAMEINUSE(prefix, params)
+            # Delay changing nickname until the user is fully signed in.
+            reactor.callLater(10, self.nickChanged, self._attemptedNick)
+        else:
+            if self.nick_change_loop.running:
+                self.nick_change_loop.stop()
+            self.msg('NickServ', 'RECOVER {}'.format(self.intended_nickname))
+            self.msg('NickServ', 'RELEASE {}'.format(self.intended_nickname))
```

Comment:
Conventionally, release is called after recover because recover puts your nickname in "custody." You then "release" the nick from custody. I'm wondering if we should instead be ghosting the nick though.

Reply:
I included this in case someone was trying to hijack a -slack nickname, but I don't think that's very likely now (since they'd need the nickserv password, and therefore would need to be staff). So I'll probably just remove this.